File: Deepspace_Java/P4/src/deepspace/SpaceCity.java
/**
 * @author BOLÍVAR PELÁEZ, CLARA
 * @author RUIZ GÓMEZ, SOLEDAD
 * 2º DGIIM - PDOO - Practice 4 DeepSpace
 */
package deepspace;

import java.util.ArrayList;
import java.util.Iterator;

/**
 *
 * @author clara
 */
public class SpaceCity extends SpaceStation {
    
    private SpaceStation base;
    private ArrayList<SpaceStation> collaborators;
    
    public SpaceCity (SpaceStation b, ArrayList<SpaceStation> c) {
        super(b);
        base = b;
        collaborators = new ArrayList<> (c);
    }
    
    
    public ArrayList<SpaceStation> getCollaborators() {
        return collaborators;
    }
    
    @Override
    public float fire () {
        
        float output = 0;
        output += base.fire();
        
        Iterator<SpaceStation> it = collaborators.iterator();
         
        while (it.hasNext()) {
            output += it.next().fire();
        }
        
        return output;
    }
    
    @Override
    public float protection () {
        
        float output = 0;
        output += base.protection();
        
        Iterator<SpaceStation> it = collaborators.iterator();
         
        while (it.hasNext()) {
            output += it.next().protection();
        }
        
        return output;
    
    }
    
    @Override
    public Transformation setLoot (Loot loot) {
    
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
        // Se hacen los mismos cálculos (no copiamos código) pero nunca se 
        // transforma la Space<city porque ya no puede más, así que siempre
        // returneamos NOTRANSFORM
    }
    
    /**
    * @brief builds a new SpaceCityToUI object from *this
    * @return WeaponToUI
    */
    @Override
    SpaceCityToUI getUIversion () {
        return new SpaceCityToUI (this);
    }
    
    /**
     * @brief toString
    * @return String with info about the instance
    */
     /*@Override
    public String toString () {
        
        String output = super.toString();
        output += "\n Los colaboradores son: \n";
        
        int size = collaborators.size();
        for(int i=0; i< size; i++)
            output += collaborators[i].toString();
            
        
        return output;
    }*/
    
  }

File: Deepspace_Ruby/P4/SpaceCity.rb
# BOLÍVAR PELÁEZ, CLARA
# RUIZ GÓMEZ, SOLEDAD
# 2º DGIIM - PDOO - Practice 4 DeepSpace

# SpaceCity
# 

module Deepspace

    class SpaceCity < SpaceStation

        attr_reader :collaborators
        
        # Constructores
        def initialize (b, c)
            newCopy(b)
            @base = b
            @collaborators = c
        end

        # fire
        def fire 
            output = @base.fire 

            collaborators.each do |coll|
                output += coll.fire
            end

            return output
        end

        # protection
        def protection 
            output = @base.protection 

            collaborators.each do |coll|
                output += coll.protection
            end

            return output
        end

        # setLoot
        def setLoot (loot)
            super
            return Transformation::NOTRANSFORM
        end

        # ToUI
        # builds a new SpaceCityToUI object from self
        def getUIversion
            SpaceCityToUI.new(self)
        end

        # to_s
        def to_s
            return getUIversion().to_s
        end



    end
end
